### PR TITLE
*: optimize Dockerfile

### PR DESCRIPTION
**Pull request by No0key**
Hello!

This PR optimizes the `Dockerfile`. 

Changes include:
- Combined RUN commands to reduce the number of layers
- Cached Go dependencies to speed up build times
- Cleaned up unnecessary apt files to reduce image size
- Consolidated user creation and permission setting commands

Reduced image size from 370MB to 302MB

category: refactor
ticket: none

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,14 +1,22 @@
...
+
 # Install dependencies
-RUN apt-get update && apt-get install -y build-essential git
+RUN apt-get update && apt-get install -y --no-install-recommends build-essential git \
+  && rm -rf /var/lib/apt/lists/*
+
 # Prep and copy source
 WORKDIR /app/charon
+COPY go.mod go.sum ./
+RUN go mod download
```
Comment (KaloyanTanev) [inline Dockerfile:11]: This is not necessary as go build will download the dependencies.

File: Dockerfile [hunk 2]
```diff
@@ -1,14 +1,22 @@
 # Container for building Go binary.
 FROM golang:1.23.1-bookworm AS builder
+
 # Install dependencies
-RUN apt-get update && apt-get install -y build-essential git
+RUN apt-get update && apt-get install -y --no-install-recommends build-essential git \
+  && rm -rf /var/lib/apt/lists/*
```
Comment (KaloyanTanev) [inline Dockerfile:6]: Not necessary, as this is multi-stage build and the image from the first stage is not used in the final image.

File: Dockerfile [hunk 3]
```diff
@@ -18,33 +26,41 @@ RUN echo "Built charon version=$(./charon version)"
...
     --no-create-home \
     --uid "$UID" \
-    "$USER"
-RUN chown charon /usr/local/bin/charon
-RUN chmod u+x /usr/local/bin/charon
+    "$USER" \
+  && chown "$USER" /usr/local/bin/charon \
+  && chmod u+x /usr/local/bin/charon \
+  && mkdir -p "/opt/$USER" \
+  && chown "$USER" "/opt/$USER"
```
Comment (KaloyanTanev) [inline Dockerfile:52]: Not necessary, as the `adduser` command already created this directory, see `--home "/opt/$USER"`. Its owner is also the said user.

File: Dockerfile (hunk 2 above)
Comment (No0key) [reply Dockerfile]: I removed the cache deletion, but now the code quality test suggests that the cache must be cleared.

**Comment by No0key** [comment]
@KaloyanTanev can you rerun tests? title name was fixed

**Comment by pinebit** [comment]
I am good with this PR, will delegate final approval and merge to @KaloyanTanev

**Review by KaloyanTanev — APPROVED** [review]
Looks good! Sorry for the delay

**Pull request merged**